### PR TITLE
Improve external types and JSON codec

### LINKED ISSUE #2167: Add static type for Cadence external `PathLink` value
### Issue To Be Solved

`cadence.PathLink` doesn't have a static type.  It is needed because:
- [CCF specs (draft)](https://github.com/fxamacker/ccf_draft) needs to include `cadence.PathLink` static type.
- CCF codec (WIP) needs all Cadence external values to have a static type.

Updates #2157 

### Suggested Solution

Maybe add `PathLinkType` and return `PathLinkType` from `PathLink.Type()`.

EDIT: Renamed `Link` to its new name `PathLink`.

### PR DESCRIPTION
**Pull request by turbolent**
Closes #2167 

## Description

- Remove unnecessary link decoding from JSON-CDC codec. Only encoding is necessary
- Add support for account links to `cadence` (add `cadence.AccountLink` and encoding in JSON-CDC)

Also: 
- Remove unnecessary/old type ID from external function type and restricted type
  (See https://dapperlabs.slack.com/archives/C034CHNNRFD/p1679610220252009. Noticed this in the new CCF codec and realized it is there for historic reasons: We used to only have the ID, but added parameters and return type later)
- Improve the way type IDs are generated in `sema` (for e.g. reference types, capability types, function types) and use those functions from `cadence`, instead of duplicating the ID generation code (and introducing a possibility for diverging logic, as e.g. attempted in [`7a1b3f5` (#2364)](https://github.com/onflow/cadence/pull/2364/commits/7a1b3f52bdd1ce36504e0e0222ec34d656ad430c#diff-b990161986c40e0867e71c60779bc98a8730a9eb0ca5866ae8189e126863c059R1853-R1867) and https://github.com/onflow/cadence/pull/2364/commits/54c69875633b42577b942a5ecbeba0b687b35c22#diff-b990161986c40e0867e71c60779bc98a8730a9eb0ca5866ae8189e126863c059R1736-R1751)
- Include missing type parameters in function type

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by fxamacker — COMMENTED** [review]
Looks good!  I mostly focused on external types, values, and JSON-CDC encoding.  

Only have one question and a few very minor nits.

**Comment by fxamacker** [comment]
@turbolent To confirm my understanding of this change's impact on CCF specs and CCF codec #2364:

- CCF doesn't need to add support for `AccountLink` and `PathLink` values.
- CCF needs to add encoding for type parameters in function type.

Is this correct?

File: types.go [hunk 1]
```diff
@@ -1819,38 +1857,41 @@ type RestrictedType struct {
...
 
 func (*RestrictedType) isType() {}
 
 func (t *RestrictedType) ID() string {
+	if t.typeID == "" {
+		var restrictionStrings []string
+		restrictionCount := len(t.Restrictions)
+		if restrictionCount > 0 {
+			restrictionStrings = make([]string, 0, restrictionCount)
+			for _, restriction := range t.Restrictions {
```
Comment (SupunS) [inline types.go:1886]: would it be better to use the `t.restrictionSet` instead of `t.Restrictions`?

**Review by SupunS — APPROVED** [review]

File: types.go [hunk 2]
```diff
@@ -1696,42 +1714,66 @@ func (t *ContractInterfaceType) Equal(other Type) bool {
 
 // TODO: type parameters
 type FunctionType struct {
-	ReturnType Type
-	typeID     string
-	Parameters []Parameter
+	TypeParameters []TypeParameter
```
[the remark this answers is not in the document]
Comment (turbolent) [reply types.go:1716]: Good point, added in f20f3ae5a3662d1ecad2d0fb7e84f3ee0684120f

File: encoding/json/encode.go [hunk 3]
```diff
@@ -643,12 +668,20 @@ func prepareFields(fieldTypes []cadence.Field, results typePreparationResults) [
 	return fields
 }
 
+func prepareTypeParameters(typeParameters []cadence.TypeParameter, results typePreparationResults) []jsonTypeParameter {
+	result := make([]jsonTypeParameter, 0)
```
[the remark this answers is not in the document]
Comment (turbolent) [reply encoding/json/encode.go]: Good idea, fixed here and elsewhere in 1d226752b40a493d2bd5e3e2e2804140b81d09ee

File: encoding/json/encode.go [hunk 4]
```diff
@@ -643,12 +668,20 @@ func prepareFields(fieldTypes []cadence.Field, results typePreparationResults) [
...
+	for _, typeParameter := range typeParameters {
+		result = append(result, prepareTypeParameter(typeParameter, results))
+	}
+	return result
+}
+
 func prepareParameters(parameterTypes []cadence.Parameter, results typePreparationResults) []jsonParameterType {
-	parameters := make([]jsonParameterType, 0)
+	result := make([]jsonParameterType, 0)
```
[the remark this answers is not in the document]
Comment (turbolent) [reply encoding/json/encode.go]: Good idea! Added pre-allocation here and elsewhere in 1d226752b40a493d2bd5e3e2e2804140b81d09ee

**Comment by turbolent** [comment]
@fxamacker 

> To confirm my understanding of this change's impact on CCF specs and CCF codec #2364:
> 
>     * CCF doesn't need to add support for `AccountLink` and `PathLink` values.
> 
>     * CCF needs to add encoding for type parameters in function type.
> 
> 
> Is this correct?

* Yes, CCF does not need support for links for now. Encoding of links to JSON is only used by the Playground for now, but will eventually be refactored, so I don' think we'll need support for them in CCF. We can add it later if needed.

* Yes, we should maybe encode type parameters for function types in CCF from the get-go

File: types.go (hunk 1 above)
Comment (turbolent) [reply types.go:1907]: No, `restrictionSet` is a `map`, so the order of iteration is non-deterministic.

Unrelated, but optimized the restriction set in 3f90c885e20d9b022764a930824c2511aa96e4ef

**Review by fxamacker — APPROVED** [review]

**Review by SupunS — APPROVED** [review]

**Pull request merged**